### PR TITLE
Redis failover feature

### PR DESCRIPTION
**Pull request by abserari**
# Description

Redis failover feature on Redis statestore component.

## Issue reference

https://github.com/dapr/components-contrib/issues/458

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Review by eduardomb08 — COMMENTED** [review]
There are some typos that need fixing.

**Comment by abserari** [comment]
Hello, I need the review help. I don't know if I had made a breaking change to binding when I do merge to resolve the conflict  and something may be wrong.

File: state/redis/redis.go [hunk 1]
```diff
@@ -99,6 +102,24 @@ func parseRedisMetadata(meta state.Metadata) (metadata, error) {
 		m.maxRetryBackoff = time.Duration(parsedVal)
 	}
 
+	m.failover = defaultFailover
+	if val, ok := meta.Properties[failover]; ok && val != "" {
+		failover, err := strconv.ParseBool(val)
+		if err != nil {
+			return m, fmt.Errorf("redis store error: can't parse failover field: %s", err)
+		}
+		m.failover = failover
```
Comment (yaron2) [inline state/redis/redis.go:111]: Probably don't need this as the default value for bool is false

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: state/redis/redis.go (hunk 1 above)
Comment (abserari) [reply state/redis/redis.go:111]: what do you mean? Remove the `m.failover = failover` and add `m.failover, err := strconv.ParseBool(val)`

Comment (yaron2) [reply state/redis/redis.go:111]: This line:

m.failover = defaultFailover

Comment (abserari) [reply state/redis/redis.go:109]: Ok now, I thought it could remind the user it's false default.

File: state/redis/redis.go [hunk 2]
```diff
@@ -109,23 +128,45 @@ func (r *StateStore) Init(metadata state.Metadata) error {
...
+		opts := &redis.Options{
+			Addr:            m.host,
+			Password:        m.password,
+			DB:              defaultDB,
+			MaxRetries:      m.maxRetries,
+			MaxRetryBackoff: m.maxRetryBackoff,
 		}
+
+		/* #nosec */
+		if m.enableTLS {
```
Comment (yaron2) [inline state/redis/redis.go:160]: Can we move this if statement to line 165 and remove the duplicate code?

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: state/redis/redis.go (hunk 2 above)
Comment (abserari) [reply state/redis/redis.go]: Move them to function now. Does this good?

File: state/redis/redis.go [hunk 3]
```diff
@@ -110,6 +129,23 @@ func (r *StateStore) Init(metadata state.Metadata) error {
...
 	r.metadata = m
 
+	if r.metadata.failover {
+		r.client = r.newFailoverClient(m)
+	} else {
+		r.client = r.newClient(m)
+	}
+
+	_, err = r.client.Ping().Result()
+	if err != nil {
```
Comment (tcnghia) [inline state/redis/redis.go:139]:
```suggestion
	if _, err = r.client.Ping().Result(); err != nil {
```

File: state/redis/redis.go [hunk 4]
```diff
@@ -125,15 +161,26 @@ func (r *StateStore) Init(metadata state.Metadata) error {
...
+	opts := &redis.FailoverOptions{
+		MasterName:      r.metadata.sentinelMasterName,
+		SentinelAddrs:   []string{r.metadata.host},
+		DB:              defaultDB,
+		MaxRetries:      m.maxRetries,
+		MaxRetryBackoff: m.maxRetryBackoff,
 	}
 
-	r.replicas, err = r.getConnectedSlaves()
+	/* #nosec */
```
Comment (tcnghia) [inline state/redis/redis.go:176]: Please elaborate on this code comment?

Comment (yaron2) [reply state/redis/redis.go:176]: That's a command to tell the linter to skip a check here

**Comment by yaron2** [comment]
@abserari Please update the Redis State Store docs in the How-To section to include the fail-over client.

https://github.com/dapr/docs/issues/791

**Pull request merged**